_posts/2022-10-01-arrow-parquet-encoding-part-3.md

```diff
@@ -0,0 +1,169 @@
+        b2: ListArray
+          Offsets: [0, 0, 2, 2, 2]
+          Validity: [false, true, false, ARBITRARY]
+          Children:
+            element: PrimitiveArray
+              Buffer[0]: [3, 4]
+```
+
+In order to encode lists, Parquet stores an integer repetition level in addition to a definition level. A repetition level identifies where in the hierarchy of repeated fields the current value is to be inserted. A value of 0 would imply a new list in the top-most repeated field, a value of 1 a new element within the top-most repeated field, a value of 2 a new element within the second top-most repeated field, and so on.
+
+Each repeated field also has a corresponding definition level, however, in this case rather than indicating a null value, they indicate an empty array.
+
+```text
+a:
+  Data Page:
+    Repetition Levels: encode([0, 0, 0, 1])
+    Definition Levels: encode([3, 0, 2, 2])
+    Values: encode([1])
+b.b1:
+  Data Page:
+    Repetition Levels: encode([0, 1, 0, 0])
+    Definition Levels: encode([2, 2, 2, 1])
+    Values: encode([1, 1, 2])
+b.b2
+  Data Page:
+    Repetition Levels: encode([0, 1, 2, 0, 0])
+    Definition Levels: encode([2, 4, 4, 2, 1])
+    Values: encode([3, 4])
+```
+
+## Additional Complications
+
+This series of posts has necessarily glossed over a number of details that further complicate matters:
+
+* A `ListArray` may contain a non-empty offset range that is masked by a validity mask
+* Reading a given number of rows from a nullable field requires reading the definition levels and determining the number of values to read based on the number of nulls present
+* Reading a given number of rows from a repeated field requires reading the repetition levels and detecting a new row based on a repetition level of 0
+* A parquet file may contain multiple row groups, each containing multiple column chunks
+* A column chunk may contain multiple pages, and there is no relationship between pages across columns
+* Parquet has alternative schema for representing lists with varying degrees of nullability
+* And more…
+
+## Summary
+As we have shown, whilst both parquet and arrow are billed as columnar formats supporting nested data, the way they represent this differs quite significantly and conversion between the two is fairly complex.
+
+Fortunately, with the Rust [parquet](https://crates.io/crates/parquet) implementation, reading and writing nested data either into Arrow or other formats is as simple as reading unnested data, with all the complex record shredding handled automatically for you. With this and other exciting features, such as out of the box support for [reading asynchronously](https://docs.rs/parquet/22.0.0/parquet/arrow/async_reader/index.html) from [object storage](https://docs.rs/object_store/0.5.0/object_store/), and advanced row filter pushdown, blog post to follow, it is the fastest and most feature complete Rust parquet implementation. We look forward to seeing what you build with it!
```

Comment:
```suggestion
Fortunately, with the Rust [parquet](https://crates.io/crates/parquet) implementation, reading and writing nested data either into Arrow is as simple as reading unnested data, with all the complex record shredding handled automatically for you. With this and other exciting features, such as out of the box support for [reading asynchronously](https://docs.rs/parquet/22.0.0/parquet/arrow/async_reader/index.html) from [object storage](https://docs.rs/object_store/0.5.0/object_store/), and advanced row filter pushdown, blog post to follow, it is the fastest and most feature complete Rust parquet implementation. We look forward to seeing what you build with it!
```
I think this might be misleading as the row APIs do not support nested data properly, the lower-level APIs do of course support it, but they don't handle the record shredding for you :sweat_smile: